Lib/test/test_unittest/testmock/testmagicmethods.py

```diff
@@ -331,6 +331,37 @@ def test_magic_methods_fspath(self):
         self.assertEqual(os.fspath(mock), expected_path)
         mock.__fspath__.assert_called_once()
 
+    def test_magic_mock_does_not_reset_magic_returns(self):
+        # https://github.com/python/cpython/issues/123934
+        for reset in (True, False):
+            with self.subTest(reset=reset):
+                mm = MagicMock()
+                self.assertIs(type(mm.__str__()), str)
+                mm.__str__.assert_called_once()
+
+                self.assertIs(type(mm.__hash__()), int)
+                mm.__hash__.assert_called_once()
+
+                for _ in range(3):
+                    # Repeat reset several times to be sure:
+                    mm.reset_mock(return_value=reset)
+
+                    self.assertIs(type(mm.__str__()), str)
+                    mm.__str__.assert_called_once()
+
+                    self.assertIs(type(mm.__hash__()), int)
+                    mm.__hash__.assert_called_once()
+
+    def test_magic_mock_resets_manual_mocks(self):
+        mm = MagicMock()
+        mm.__iter__ = MagicMock(return_value=iter([1]))
```

Comment:
Hmm, okay, things I didn't know about MagicMock and I'm not sure I like ;-) :

```python
>>> m = MagicMock()
>>> list(iter(m))
[]
```

However, what happens if you change the above line to:

```suggestion
        mm.__iter__ = MagicMock(return_value=[])
```

I'm not sure it makes any difference, but mainly 'cos of the weird behaviour I started with an example of ;-) 

Here's some playing around in a Python 3.12 interpreter that might give more of a feel for what I'm trying to communicate:

```python
>>> from unittest.mock import MagicMock
>>> m = MagicMock()
>>> m.__iter__
<MagicMock name='mock.__iter__' id='4307758752'>
>>> m.__iter__.return_value = []
>>> iter(m)
<list_iterator object at 0x10088f8b0>
>>> list(iter(m))
[]
>>> list(iter(m)) is m.__iter__.return_value
False
>>> m.__iter__.return_value = [1, 2, 3]
>>> list(iter(m))
[1, 2, 3]
```

Reply:
I added one more test with the behavior you are showing here. I executed these two tests

```python
    def test_magic_mock_resets_manual_mocks(self):
        mm = MagicMock()
        mm.__iter__ = MagicMock(return_value=iter([1]))
        mm.custom = MagicMock(return_value=2)
        self.assertEqual(list(iter(mm)), [1])
        self.assertEqual(mm.custom(), 2)

        mm.reset_mock(return_value=True)
        self.assertEqual(list(iter(mm)), [])
        self.assertIsInstance(mm.custom(), MagicMock)

    def test_magic_mock_resets_manual_mocks_empty_iter(self):
        mm = MagicMock()
        mm.__iter__.return_value = []
        self.assertEqual(list(iter(mm)), [])

        mm.reset_mock(return_value=True)
        self.assertEqual(list(iter(mm)), [])
```

on `main` (with no changes from this PR) and it was also successful.

Should I cover anything else?

Reply:
Ah, I think this is where I got confused; the docs aren't great here. I believe the intention may be that `return_value` is _only_ a boolean, or it may be that it's intended as a way to provide a new value.

If you feel inclined, it'd be good to clarify [the docs](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.reset_mock) on this.

Reply:
Sure!